File: frontend/src/types/case.ts
import { User } from "./user";
import { Document } from "./document";
import { Communication } from "./communication";

export enum CaseStatus {
  OPEN = "OPEN",
  ACTIVE = "ACTIVE",
  UNDER_REVIEW = "UNDER_REVIEW",
  COMPLETED = "COMPLETED",
  ON_HOLD = "ON_HOLD",
  CLOSED = "CLOSED",
}

export enum CasePriority {
  LOW = "LOW",
  MEDIUM = "MEDIUM",
  HIGH = "HIGH",
  URGENT = "URGENT",
}

export enum CaseCategory {
  GENERAL = "GENERAL",
  LEGAL = "LEGAL",
  FINANCIAL = "FINANCIAL",
  HR = "HR",
  COMPLIANCE = "COMPLIANCE",
  OPERATIONS = "OPERATIONS",
  STRATEGIC = "STRATEGIC",
}

export interface Case {
  id: string;
  title: string;
  description: string;
  status: CaseStatus;
  priority: CasePriority;
  category: CaseCategory;
  tags: string[];
  assignedToId?: string;
  assignedTo?: User;
  createdById: string;
  createdBy?: User;
  documents?: Document[];
  communications?: Communication[];
  dueDate?: string;
  estimatedCompletionDate?: string;
  actualCompletionDate?: string;
  budget?: number;
  actualCost?: number;
  location?: string;
  department?: string;
  stakeholderIds?: string[];
  stakeholders?: User[];
  createdAt: string;
  updatedAt: string;
}

export interface CreateCaseRequest {
  title: string;
  description: string;
  status?: CaseStatus;
  priority?: CasePriority;
  category?: CaseCategory;
  tags?: string[];
  assignedToId?: string;
  dueDate?: string;
  estimatedCompletionDate?: string;
  budget?: number;
  location?: string;
  department?: string;
  stakeholderIds?: string[];
  attachments?: File[];
}

export interface UpdateCaseRequest {
  title?: string;
  description?: string;
  status?: CaseStatus;
  priority?: CasePriority;
  category?: CaseCategory;
  tags?: string[];
  assignedToId?: string;
  dueDate?: string;
  estimatedCompletionDate?: string;
  actualCompletionDate?: string;
  budget?: number;
  actualCost?: number;
  location?: string;
  department?: string;
  stakeholderIds?: string[];
}

File: backend/src/main/java/com/efile/core/casemanagement/Case.java
package com.efile.core.casemanagement;

import com.efile.core.user.User;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.time.Instant;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;
import lombok.AllArgsConstructor;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;

@Entity
@Table(
    name = "cases",
    indexes = {
        @jakarta.persistence.Index(name = "idx_cases_assigned_to", columnList = "assigned_to_id"),
        @jakarta.persistence.Index(name = "idx_cases_status", columnList = "status"),
        @jakarta.persistence.Index(name = "idx_cases_priority", columnList = "priority"),
        @jakarta.persistence.Index(name = "idx_cases_category", columnList = "category"),
        @jakarta.persistence.Index(name = "idx_cases_created_at", columnList = "created_at")
    }
)
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(of = "id")
@ToString(of = "id")
public class Case {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 191)
    private String title;

    @Column(nullable = false, columnDefinition = "text")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 32)
    private CaseStatus status = CaseStatus.OPEN;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 32)
    private CasePriority priority = CasePriority.MEDIUM;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 32)
    private CaseCategory category = CaseCategory.GENERAL;

    @Column(columnDefinition = "json")
    private String tags; // Store as JSON string

    @Column(name = "due_date")
    private Instant dueDate;

    @Column(name = "estimated_completion_date")
    private Instant estimatedCompletionDate;

    @Column(name = "budget")
    private Double budget;

    @Column(length = 255)
    private String location;

    @Column(length = 100)
    private String department;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assigned_to_id")
    private User assignedTo;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by_id", nullable = false)
    private User createdBy;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private Instant updatedAt;
}
